server/modules/module_state_machine.js

```diff
@@ -38,285 +40,201 @@ function isDisplayInPoly(rect, poly) {
+    // we have no def for it.
+    // TODO(applmak): Handle this better.
+    this.state.playModule(moduleName);
   }
   getGeo() {
-    return this.geo_;
+    return this.context_.geo;
   }
   getDeadlineOfNextTransition() {
-    return this.current_.getDeadline();
-  }
-  stop() {
-    this.context_.server.stop();
+    return this.state.getDeadline();
   }
 }
 
 class IdleState extends stateMachine.State {
-  constructor() {
-    super('IdleState');
-  }
-  enter_() {}
-  loadPlaylist(deadline) {
-    this.transition_(new LoadingState(deadline));
+  enter(transition) {
+    this.transition_ = transition;
   }
   newClient(client) {}
   playModule(moduleName) {
-    return false;
+    throw new Error(`No module named ${moduleName}`);
   }
   getDeadline() {
     return Infinity;
   }
 }
 
-class LoadingState extends stateMachine.State {
-  constructor(deadline) {
-    super('LoadingState');
-
-    this.indexOfModuleToPlayWhenLoaded_ = 0;
-
-    // The time at which we should swap to the new playlist.
-    this.deadline_ = deadline;
-  }
-  enter_() {
-    // Load the playlist.
-    var modulePromises = this.context_.playlist.map((m) => m.whenLoadedPromise);
-    Promise.allSettled(modulePromises).done((results) => {
-      // Check to see if any modules were rejected. If so, remove them from the 
-      // playlist.
-      this.context_.playlist = this.context_.playlist.filter((module, index) => {
-        if (results[index].status != 'resolved') {
-          debug('Removing ' + module + ' from the playlist!');
-        }
-        return results[index].status == 'resolved';
-      });
-      this.transition_(new TransitionState(this.deadline_, this.indexOfModuleToPlayWhenLoaded_));
-    });
-  }
-  loadPlaylist(deadline) {
-    // Stop any prior async work from continuing.
-    // The fact that we still be in the 'map' above is no big deal, we'll just
-    // extra disk i/o.
-    this.transition_(new LoadingState(deadline));
-  }
-  newClient(client) {}
-  playModule(moduleName) {
-    var index = _.findIndex(this.context_.playlist, (d) => d.name == moduleName);
-    if (index != -1) {
-      debug('Will play module ' + moduleName + ' when loading is complete.');
-      this.indexOfModuleToPlayWhenLoaded_ = index;
-      return true;
-    }
-    debug('Could not find requested module ' + moduleName);
-    return false;
-  }
-  getDeadline() {
-    return this.deadline_;
-  }
-}
-
-class TransitionState extends stateMachine.State {
-  constructor(deadline, opt_index) {
-    super('TransitionState');
+class DisplayState extends stateMachine.State {
+  constructor(playlist, layout, deadline, index = 0) {
+    super();
 
-    // If no next item passed, choose the first one.
-    this.index_ = opt_index || 0;
+    this.layout_ = layout;
+    this.playlist_ = playlist;
+    
+    // Index cannot be out of bounds.
+    assert(index < playlist.length);
+    this.index_ = index;
 
+    this.displayDuration_ = this.layout_.moduleDuration * 1000;
```

Comment:
it would be nice after this to go through and name the time variables with units in the name.